File: Design Browser History.cpp
class BrowserHistory {
    
//     Node doubly linked list
   class Node{
        public:
        string url;
        Node* next = NULL;
        Node* prev = NULL;
     };
    
     Node* head = NULL;
     Node* tail = NULL;
     Node* current = NULL;
public:
    BrowserHistory(string homepage) {
        Node* newNode = new Node;
        newNode->url = homepage;
        head = tail = current = newNode;
    }
    
    void visit(string url) {
        Node* tempCurrent = current->next;
        while(tempCurrent != NULL){
            Node* tempHold = tempCurrent->next;
            delete tempCurrent;
            tempCurrent = tempHold;
        }
        current->next = new Node;
        current->next->url = url;
        current->next->prev = current;
        current = current->next;
        tail = current;
    }
    
    string back(int steps) {
        int x = 0;
        while(current != head && x < steps){
            current = current->prev;
            x++;
        }
        return current->url;
    }
    
    string forward(int steps) {
        int x = 0;
        while(current != tail && x < steps){
            current = current->next;
            x++;
        }
        return current->url;
    }
};

File: Design Browser History.java
class BrowserHistory {
    private List<String> history;
    private int currentIndex;
    
    public BrowserHistory(String homepage) {
        history = new ArrayList<>();
        history.add(homepage);
        currentIndex = 0;
    }
    
    public void visit(String url) {
        // Remove all the URLs after the current index
        int n = history.size();
        for (int i = n - 1; i > currentIndex; i--) {
            history.remove(i);
        }
        // Add the new URL to the end of the history
        history.add(url);
        currentIndex = history.size() - 1;
    }
    
    public String back(int steps) {
        currentIndex = Math.max(currentIndex - steps, 0);
        return history.get(currentIndex);
    }
    
    public String forward(int steps) {
        int n = history.size();
        currentIndex = Math.min(currentIndex + steps, n - 1);
        return history.get(currentIndex);
    }
}
